typing/typecore.ml

```diff
@@ -3202,6 +3184,70 @@ and type_expect_
                      sexp.pexp_attributes) ::
                       exp.exp_extra;
       }
+  | Pexp_letop{ let_ = slet; ands = sands; body = sbody } ->
+      let rec loop spat_acc ty_acc sands =
+        match sands with
+        | [] -> spat_acc, ty_acc
+        | { pbop_pat = spat; _} :: rest ->
+            let ty = newvar () in
+            let loc = { slet.pbop_op.loc with Location.loc_ghost = true } in
+            let spat_acc = Ast_helper.Pat.tuple ~loc [spat_acc; spat] in
+            let ty_acc = newty (Ttuple [ty_acc; ty]) in
+            loop spat_acc ty_acc rest
+      in
+      if !Clflags.principal then begin_def ();
+      let let_loc = slet.pbop_op.loc in
+      let op_path, op_desc = type_binding_op_ident env slet.pbop_op in
+      let op_type = instance op_desc.val_type in
+      let spat_params, ty_params = loop slet.pbop_pat (newvar ()) sands in
+      let ty_func_result = newvar () in
+      let ty_func = newty (Tarrow(Nolabel, ty_params, ty_func_result, Cok)) in
+      let ty_result = newvar () in
+      let ty_andops = newvar () in
+      let ty_op =
+        newty (Tarrow(Nolabel, ty_andops,
+          newty (Tarrow(Nolabel, ty_func, ty_result, Cok)), Cok))
+      in
+      begin try
+        unify env op_type ty_op
+      with Unify trace ->
+        raise(Error(let_loc, env, Letop_type_clash(slet.pbop_op.txt, trace)))
+      end;
+      if !Clflags.principal then begin
+        end_def ();
+        generalize_structure ty_andops;
+        generalize_structure ty_params;
+        generalize_structure ty_func_result;
+        generalize_structure ty_result
+      end;
+      let exp, ands = type_andops env slet.pbop_exp sands ty_andops in
+      let scase = Ast_helper.Exp.case spat_params sbody in
+      let cases, partial =
+        type_cases env ty_params ty_func_result true loc [scase]
```

Comment:
I don't really like the way you're reusing `type_cases` here.
In particular, it means that if you get an error message from parmatch, it will be about the "final" pattern, instead of the part which is actually problematic.
For example:
```
# let (let+) x f = f x                                                                                                                  
  let (and+) x y = x, y;;                                                                                                               
val ( let+ ) : 'a -> ('a -> 'b) -> 'b = <fun>
val ( and+ ) : 'a -> 'b -> 'a * 'b = <fun>
# let+ None = None and+ _ = None in ();;                                                                                                
Line 1, characters 0-36:
1 | let+ None = None and+ _ = None in ();;
    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
Warning 8: this pattern-matching is not exhaustive.
Here is an example of a case that is not matched:
(Some _, _)
- : unit = ()
```

Reply:
Note that it also has some implications in terms of typing, which I'm not sure I like:
```
# type _ t = C : int t;;
type _ t = C : int t

# let f (type a) (x : a t) (y : a) = let+ C = x and+ 3 = y in ();;
val f : 'a t -> 'a -> unit = <fun>
# let g (type a) (x : a t) (y : a) = let+ 3 = y and+ C = x in ();;
Line 1, characters 40-41:
1 | let g (type a) (x : a t) (y : a) = let+ 3 = y and+ C = x in ();;
                                            ^
Error: This pattern matches values of type int
       but a pattern was expected which matches values of type a

# #principal true;;

# let f (type a) (x : a t) (y : a) = let+ C = x and+ 3 = y in ();; 
Line 1, characters 35-39:
1 | let f (type a) (x : a t) (y : a) = let+ C = x and+ 3 = y in ();;
                                       ^^^^
Error: This pattern matches values of type int t * int
       but a pattern was expected which matches values of type a t * a
       Type int is not compatible with type a 

# let g (type a) (x : a t) (y : a) = let+ 3 = y and+ C = x in ();;
Line 1, characters 35-39:
1 | let g (type a) (x : a t) (y : a) = let+ 3 = y and+ C = x in ();;
                                       ^^^^
Error: This pattern matches values of type int * int t
       but a pattern was expected which matches values of type a * a t
       Type int is not compatible with type a 
```

Also, the error message when `-principal` is turned on is fairly distressing.

Reply:
I agree that it is not ideal. It would be quite a bit of work to divide type_cases into reasonable sub-components that can be used directly to get better error messages here. I'm not too worried about the ones that only arise in `-principal` since people don't use that and certainly beginners don't. The exhaustivity one is a bit more worrying, but I think that non-exhaustive matches in lets are pretty rare and whilst the message is a bit confusing it is not indecipherable. So I think it is fine to merge this as is and work on improving the error messages later.

Note that the support for GADTs is deliberate and tested in the test suite. The `and+` here is the monoidal operation of an applicative/monad which is associative but not generally commutative -- so there is a meaningful order between the patterns and it makes sense to allow equations between them.

Reply:
> Note that the support for GADTs is deliberate and tested in the test suite. The and+ here is the monoidal operation of an applicative/monad which is associative but not generally commutative -- so there is a meaningful order between the patterns and it makes sense to allow equations between them.

Ack.

Reply:
As for the issues regarding the error messages, I would personally prefer to delay inclusion of the feature until a "proper" implementation is done.
At the same time:
- I don't know how long it would take for such an implementation to appear
- it's not the first time that unpolished things would get in
- I have the impression that various people are very keen on having the feature sooner rather than later.

So instead of just blocking this feature, I'm happy to leave the inclusion decision to someone else if/when it reduces to just that point.